src/libraries/System.Text.Json/tests/System.Text.Json.Tests/Serialization/MetadataTests/DefaultJsonTypeInfoResolverTests.JsonPropertyInfo.cs

```diff
@@ -0,0 +1,1048 @@
+            {
+                if (ti.Type == typeof(TestClassWithNumber))
+                {
+                    Assert.Null(ti.Properties[0].ShouldSerialize);
+                    ti.Properties[0].ShouldSerialize = (o, val) =>
+                    {
+                        Assert.Same(obj, o);
+                        int intValue = (int)val;
+                        Assert.Equal(obj.IntProperty, intValue);
+                        return intValue != 3;
+                    };
+                }
+            });
+
+            JsonSerializerOptions o = new();
+            o.DefaultIgnoreCondition = JsonIgnoreCondition.WhenWritingDefault;
+            o.TypeInfoResolver = resolver;
+
+            string json = JsonSerializer.Serialize(obj, o);
+            Assert.Equal("""{"IntProperty":37}""", json);
+
+            obj.IntProperty = default;
+            json = JsonSerializer.Serialize(obj, o);
+            Assert.Equal("""{"IntProperty":0}""", json);
+
+            obj.IntProperty = 3;
+            json = JsonSerializer.Serialize(obj, o);
+            Assert.Equal("{}", json);
+        }
+
+        [Fact]
+        public static void DefaultIgnoreConditionFromOptionsIsRespectedWhenShouldSerializePropertIsAssignedAndCleared()
+        {
+            TestClassWithNumberAndIgnoreConditionOnProperty obj = new()
+            {
+                IntProperty = 37,
+            };
+
+            DefaultJsonTypeInfoResolver resolver = new();
+            resolver.Modifiers.Add((ti) =>
+            {
+                if (ti.Type == typeof(TestClassWithNumber))
+                {
+                    Assert.Null(ti.Properties[0].ShouldSerialize);
+                    ti.Properties[0].ShouldSerialize = (o, val) =>
+                    {
+                        Assert.Same(obj, o);
+                        int intValue = (int)val;
+                        Assert.Equal(obj.IntProperty, intValue);
+                        return intValue != 3;
+                    };
+
+                    ti.Properties[0].ShouldSerialize = null;
+                }
+            });
+
+            JsonSerializerOptions o = new();
+            o.DefaultIgnoreCondition = JsonIgnoreCondition.WhenWritingDefault;
+            o.TypeInfoResolver = resolver;
+
+            string json = JsonSerializer.Serialize(obj, o);
+            Assert.Equal("""{"IntProperty":37}""", json);
+
+            obj.IntProperty = default;
+            json = JsonSerializer.Serialize(obj, o);
+            Assert.Equal("{}", json);
+
+            obj.IntProperty = 3;
+            json = JsonSerializer.Serialize(obj, o);
+            Assert.Equal("""{"IntProperty":3}""", json);
+        }
+
+        public enum ModifyJsonIgnore
+        {
+            DontModify,
+            NeverSerialize,
+            AlwaysSerialize,
+            DontSerializeNumber3OrStringAsd,
+        }
+
+        [Theory]
+        [InlineData(JsonIgnoreCondition.WhenWritingDefault, ModifyJsonIgnore.DontModify)]
```

Comment:
So we have any unit tests validating that the `JsonIgnoreCondition` to `ShouldSerialize` delegate behaves as expected? (outside of serialization I mean)

Reply:
See TestJsonIgnoreConditionDelegate